range-requests/src/lib.rs

```diff
@@ -0,0 +1,530 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use dropshot::Body;
+use dropshot::HttpError;
+use futures::TryStreamExt;
+use http::HeaderValue;
+use hyper::{
+    header::{ACCEPT_RANGES, CONTENT_LENGTH, CONTENT_RANGE, CONTENT_TYPE},
+    Response, StatusCode,
+};
+
+const ACCEPT_RANGES_BYTES: http::HeaderValue =
+    http::HeaderValue::from_static("bytes");
+const CONTENT_TYPE_OCTET_STREAM: http::HeaderValue =
+    http::HeaderValue::from_static("application/octet-stream");
+
+/// Errors which may be returned when processing range requests
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Using multiple ranges is not supported")]
+    MultipleRangesUnsupported,
+
+    #[error("Range would overflow (start + length is too large)")]
+    RangeOverflow,
+
+    #[error("Range would underflow (total content length < start)")]
+    RangeUnderflow,
+
+    #[error("Empty Range")]
+    EmptyRange,
+
+    #[error("Failed to parse range: {0:?}")]
+    Parse(http_range::HttpRangeParseError),
+
+    #[error(transparent)]
+    Http(#[from] http::Error),
+}
+
+impl From<Error> for HttpError {
+    fn from(err: Error) -> Self {
+        match err {
+            Error::MultipleRangesUnsupported => {
+                HttpError::for_bad_request(None, err.to_string())
+            }
+            Error::RangeUnderflow
+            | Error::RangeOverflow
+            | Error::EmptyRange => HttpError::for_client_error(
+                None,
+                http::StatusCode::RANGE_NOT_SATISFIABLE,
```

Comment:
Per MDN, when returning a 419, the server is also supposed to add a `Content-Range` header indicating the actual length of the resource:

> The 416 response message should contain a Content-Range indicating an unsatisfied range (that is a '*') followed by a '/' and the current length of the resource, e.g., Content-Range: bytes */12777

(see: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/416)

I don't think it's practical to actually do that in this PR, since Dropshot doesn't presently have a nice way to add headers to error responses (see oxidecomputer/dropshot#801, oxidecomputer/dropshot#39). But, it might be worth adding a comment or something to remember to return such a header if/when Dropshot adds a way to do so. We might also want to go ahead and add a field for the actual length of the resource to these error variants, so that it can be used to generate the response headers later — for now, we could also include the actual length in the formatted representation of the errors?

Reply:
(I think I might go take a crack at implementing oxidecomputer/dropshot#39 in the near future, because I've also needed it for very different reasons and it seems like it could be fun to do...)

Reply:
I *think* we're doing this in this PR -- see `bad_range_response`, where we do attach these headers. If we try to parse a `SingleRange` from a `PotentialRange` and fail with this error, we'll attach the necessary headers.

I refactored this in 6c2562d to make this harder to mis-use - I've removed the conversion to HttpError that automatically happens, and ensured that the headers get supplied in-line.

That said, we'd still greatly benefit from https://github.com/oxidecomputer/dropshot/issues/39. I'm returning a `Response<Body>` directly, just to be able to supply these headers, and I'd really prefer not to do that.